src/lib/fetcher/fetcher.ts

```diff
@@ -0,0 +1,139 @@
+import { IRequestWrapper } from '../../dex-helper';
+import { Logger } from 'log4js';
+import { RequestConfig, Response } from '../../dex-helper/irequest-wrapper';
+
+import Timeout = NodeJS.Timeout;
+
+const FETCH_TIMEOUT_MS = 10 * 1000;
+const FETCH_FAIL_MAX_ATTEMPT = 5;
+const FETCH_FAIL_RETRY_TIMEOUT_MS = 60 * 1000;
+
+export type RequestInfo<T> = {
+  requestOptions: RequestConfig;
+  caster: (data: unknown) => T | null;
+  authenticate?: (options: RequestConfig) => void;
+  excludedFieldsCaching?: string[];
+};
+
+export type RequestInfoWithHandler<T> = {
+  info: RequestInfo<T>;
+  handler: (data: T) => void;
+};
+
+export default class Fetcher<T> {
+  private intervalId?: Timeout;
+  private requests: RequestInfoWithHandler<T>[];
+  public lastFetchSucceeded: boolean = false;
+  private failedCount = 0;
+  private stop: boolean = true;
+
+  constructor(
+    private requestWrapper: IRequestWrapper,
+    requestsInfo: RequestInfoWithHandler<T> | RequestInfoWithHandler<T>[],
+    private pollIntervalMs: number,
+    private logger: Logger,
+  ) {
+    if (Array.isArray(requestsInfo)) {
+      this.requests = requestsInfo;
+    } else {
+      this.requests = [requestsInfo];
+    }
+  }
+
+  async fetch() {
+    if (this.stop === true) {
+      return;
+    }
+
+    if (this.failedCount >= FETCH_FAIL_MAX_ATTEMPT) {
+      this.stopPolling();
+
+      setTimeout(() => {
+        this.startPolling();
+        this.failedCount = 0;
+      }, FETCH_FAIL_RETRY_TIMEOUT_MS);
+      return;
+    }
+
+    const promises = this.requests.map<Promise<Error | Response<T>>>(
+      async (reqInfo: RequestInfoWithHandler<T>) => {
+        const info = reqInfo.info;
+        let options = info.requestOptions;
+        if (info.authenticate) {
+          info.authenticate(options);
+        }
+
+        try {
+          const result = await this.requestWrapper.request({
+            timeout: FETCH_TIMEOUT_MS,
+            ...options,
+          });
+          return result;
+        } catch (e) {
+          return e as Error;
+        }
+      },
+    );
+    const results = await Promise.all(promises);
+    const failures = results
+      .map((_, i) => i)
+      .filter(i => results[i] instanceof Error);
+
+    failures.forEach(i => {
+      this.logger.warn(
+        `failed polling ${this.requests[i].info.requestOptions.url} ${results[i]}`,
+      );
+    });
+
+    if (failures.length === results.length) {
+      this.lastFetchSucceeded = false;
+      this.failedCount += 1;
+    } else {
+      this.lastFetchSucceeded = true;
+    }
+
+    results
+      .map((_, i) => i)
+      .filter(i => !(results[i] instanceof Error))
```

Comment:
You don't need to loop twice through all results. But if we expect that `requests.length` is not big, then maybe not worth optimizing

Reply:
Yes I know but I think we don't care it's more simple to read